dialogue-core/src/main/java/com/palantir/dialogue/core/HostIndexResponseMarkingChannel.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.futures.DialogueFutures;
+
+final class HostIndexResponseMarkingChannel implements Channel {
+
+    private final Integer hostIndex;
+    private final Channel delegate;
+
+    HostIndexResponseMarkingChannel(int hostIndex, Channel delegate) {
+        this.hostIndex = hostIndex;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return DialogueFutures.transform(delegate.execute(endpoint, request), this::addHostIndex);
+    }
+
+    Response addHostIndex(Response response) {
+        response.attachments().put(RoutingAttachments.HOST_KEY, hostIndex);
```

Comment:
Should we use a cooperative design here, in which the request must provide an attachment to opt into getting a response attachment describing the host?

Reply:
I don't think this would work: the first request in a sticky channel does not know the host to route to, but wants to know where the request went.

Reply:
I don't quite understand the concern here though? is it about allocating the attachments map? I also don't mind doing what you propose, IF we don't need to support per-host.